Added Data Integrity checks for TB Encounters

moshonk opened the pull request:
@ssmusoke ,

Kindly review: Added Data Integrity checks for TB Encounters

cc @ningosi

ssmusoke commented:
@moshonk please use HQL queries like the rest of the data validation rules. We don't need the additional classes which make the implementation more complex

moshonk commented:
@ssmusoke,

Kindly review this. I have refactored the rules from a patient-calculation based to a hql-query based method of data retrieval.

cc @ningosi

ssmusoke commented on api/pom.xml:
```diff
@@ -182,6 +183,11 @@
 			<artifactId>uiframework-api</artifactId>
 		</dependency>
 
+		<dependency>
```
@moshonk This is not needed if you are using HQL

moshonk replied:
Noted and removed

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,200 @@
+			
+			if (!identifierFound) {
+				RuleResult<Patient> ruleResult = new RuleResult<>();
+				ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getUuid());
```
@moshonk Here link to the Summary page which does not have a TB identifier - you can get it from the encounterId since it will be the only one in the program

moshonk replied:
Noted and addressed.

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,200 @@
+		List<RuleResult<Patient>> ruleResults = new ArrayList<>();
+		for (Patient patient : uniquePatientList) {
+			RuleResult<Patient> ruleResult = new RuleResult<>();
+			ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getUuid());
```
@moshonk Link to the Summary page for the encounter with no final outcome

moshonk replied:
Noted and addressed.

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/metadata/concept/Dictionary.java:
```diff
@@ -0,0 +1,157 @@
+package org.openmrs.module.aijar.metadata.concept;
```
@moshonk This file is not needed if you are using HQL

moshonk replied:
@ssmusoke, I lifted the whole file off KenyaEMR, it has some pretty useful methods, and I am using the getConcept() method in a number of places

ssmusoke replied:
@moshonk We do not need the concept classes anywhere - you can use the ConceptService instead

moshonk replied:
@ssmusoke, Noted

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,200 @@
+	public List<RuleResult<Patient>> evaluate() {
+		List<RuleResult<Patient>> ruleResults = new ArrayList<>();
+		
+		ruleResults.addAll(patientsWithMissingTBNumberIdentifiers("304df0d0-afe4-4a61-a917-d684b100a65a", "Unit TB Number"));
```
@moshonk Name the methods properly - patientsWithMissingUnitTBNumber, patientsWithMissingHSDTBNumber, patientsWithMissingDistrictTBNumber ... applies to duplicates also

moshonk replied:
@ssmusoke, I did this on purpose to avoid having many copy paste methods that do the same thing. The intention was to have one method that can be reused for validating the various Identifier Types.

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,200 @@
+		ruleResults.addAll(patientsWithMissingTBNumberIdentifiers("d1cda288-4853-4450-afbc-76bd4e65ea70", "HSD TB Number"));
+		ruleResults.addAll(patientsWithMissingTBNumberIdentifiers("67e9ec2f-4c72-408b-8122-3706909d77ec", "District TB Number"));
+		ruleResults.addAll(patientsWithSimilarTBNumberIdentifiersDuplicatedForASinglePatient("67e9ec2f-4c72-408b-8122-3706909d77ec", "District TB Number"));
+		ruleResults.addAll(patientsWithSimilarTBNumberIdentifiersDuplicatedForASinglePatient("304df0d0-afe4-4a61-a917-d684b100a65a", "Unit TB Number"));
```
@moshonk This method is not clear - please elaborate, you cannot have similarTBIdentifiers that are duplicated. You should be looking for:

1. Single patient with duplicate numbers across multiple treatment programs 

2. Multiple patients with the same numbers

ssmusoke replied:
@moshonk This method naming needs to be worked on

moshonk replied:
@ssmusoke, noted,

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,200 @@
+		for (Patient patient : uniquePatientList) {
+			RuleResult<Patient> ruleResult = new RuleResult<>();
+			ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getUuid());
+			ruleResult.setNotes("Patient #" + patient.getId() + " with similar " + identifierTitle + " identifiers: duplicated for a single patient,");
```
@moshonk Also add the value of the identifier here for follow up

moshonk replied:
Noted and addressed.

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,200 @@
+		for (Patient patient : uniquePatientList) {
+			RuleResult<Patient> ruleResult = new RuleResult<>();
+			ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getUuid());
+			ruleResult.setNotes("Patient #" + patient.getId() + " with similar " + identifierTitle + " identifiers: duplicated for a single patient,");
```
@moshonk Add the actual duplicated identifier value

moshonk replied:
Noted and addressed

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,200 @@
+		for (Patient patient : uniquePatientList) {
+			RuleResult<Patient> ruleResult = new RuleResult<>();
+			ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getUuid());
+			ruleResult.setNotes("Patient #" + patient.getId() + " has no final TB Outcome 9 months after start of treatment");
```
@moshonk one of the TB identifiers (if not available) then fall back to the OpenMRS ID for the patient as this is searchable not the numeric database ID

moshonk replied:
@ssmusoke, what if we use exclusively use te OpenMRS Id? Using the TB Identifiers is bound to get mucky since we are fetching the identifier values from obs. It could also greatly affect performance in the long run where we have a large number obs recorded.

ssmusoke replied:
@moshonk The OpenMRS ID is not used anywhere within UgandaEMR so we cannot fall back on it. Unfortunately we have to take the performance hit to make it useable.

moshonk replied:
@ssmusoke, noted.

ssmusoke commented on api/src/test/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncountersTest.java:
```diff
@@ -0,0 +1,64 @@
+	}
+	
+	@Test
+	public void testInvalidTBEncounters() {
```
@moshonk Let this test name match the content

moshonk replied:
Noted and addressed.

moshonk commented:
@ssmusoke, this has been noted and rectified.

ssmusoke commented:
@moshonk Please can you respond to each comment that I made - that way you are sure every area has been touched

moshonk commented:
@ssmusoke, for some reason I had only seen one comment.. my bad! I'll review and respond to the other comments

moshonk commented:
@ssmusoke, I have made the relevant changes. Here's a highlight:

- Added link to the summary page for validation rule results
- Removed the file Dictionary.java
- Show TB Number in the rule results ActionUrl
- For duplicated TB Numbers, show the actual duplicated value in the Notes

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/BasePatientRuleDefinition.java:
```diff
@@ -41,6 +49,39 @@ public String getExposedInfantNumber(Patient patient) {
+	 * @param patient
+	 * @return
+	 */
+	public String getTbNumber(Patient patient) { //To be done
```
A patient can have multiple TB numbers so you need to pass some more information here to get the TB number - maybe a date or the TB Summary page encounter not go through Programs

ssmusoke replied:
@moshonk  You do not need the most recent number but rather the TB number associated with the summary encounter.. If a patient has 3 different TB treatments then you need to show the correct identifiers for encounter being shown .. You should tie to an encounter since there is only one identifier per encounter

moshonk replied:
Ok. Noted. I'll make the necessary changes and send you the update this evening.

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/metadata/concept/Concepts.java:
```diff
@@ -0,0 +1,12 @@
+package org.openmrs.module.aijar.metadata.concept;
```
No need for this class use the Aijar Constants class

moshonk replied:
@ssmusoke, Noted and executed.

ssmusoke commented on api/pom.xml:
```diff
@@ -165,6 +165,7 @@
 		<dependency>
 			<groupId>org.openmrs.module</groupId>
 			<artifactId>reporting-api</artifactId>
+			<scope>provided</scope>
```
@moshonk This dependency is not needed

ssmusoke replied:
@moshonk Why add this dependency?

ssmusoke replied:
@moshonk This is still pending closure

moshonk replied:
@ssmusoke, Noted and removed

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/BasePatientRuleDefinition.java:
```diff
@@ -18,32 +23,75 @@
 		PatientService patientService = Context.getPatientService();
-		PatientIdentifierType pit = patientService.getPatientIdentifierTypeByUuid(PatientIdentifierTypes.HIV_CARE_NUMBER.uuid());
-		 return patient.getPatientIdentifier(pit) == null ? "" : patient.getPatientIdentifier(pit).toString();
+		PatientIdentifierType pit = patientService
```
@moshonk Why this change

moshonk replied:
@ssmusoke This happened when I ran a code format on the IDE

moshonk replied:
I have reverted it.

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/BasePatientRuleDefinition.java:
```diff
@@ -18,32 +23,75 @@
 	public String getExposedInfantNumber(Patient patient) {
 		PatientService patientService = Context.getPatientService();
-		PatientIdentifierType pit = patientService.getPatientIdentifierTypeByUuid(PatientIdentifierTypes.EXPOSED_INFANT_NUMBER.uuid());
+		PatientIdentifierType pit = patientService
```
@moshonk Why this change?

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,210 @@
+			
+			RuleResult<Patient> ruleResult = new RuleResult<>();
+			ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getUuid());
+			ruleResult.setNotes("Patient #" + getTbNumber(patient, obs.getEncounter()) + " with similar " + identifierTitle + " identifiers(" + obs.getValueText() + "): duplicated for a single patient,");
```
Correct the label of this message: from "Patient # with similar District TB Number identifiers(D-1234): duplicated for a single patient,"

to "The District TB Number (D-1234) is used by another patient"

moshonk replied:
Notes and addressed.

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,210 @@
+			Patient patient = obs.getEncounter().getPatient();
+			RuleResult<Patient> ruleResult = new RuleResult<>();
+			ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getUuid());
+			ruleResult.setNotes("Patient #" + getTbNumber(patient, obs.getEncounter()) + " with similar " + identifierTitle + " identifiers(" + obs.getValueText() + "): duplicated for a single patient,");
```
Correct the label of this message: from "Patient # with similar District TB Number identifiers(D-1234): duplicated for a single patient,"

to "The District TB Number (D-1234) is used by another patient"

moshonk replied:
Noted and addressed.

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,210 @@
+			for (Encounter encounter : patientEncounters) {
+				RuleResult<Patient> ruleResult = new RuleResult<>();
+				ruleResult.setActionUrl("htmlformentryui/htmlform/editHtmlFormWithStandardUi.page?patientId=" + patient.getUuid() + "&encounterId=" + encounter.getId());
+				ruleResult.setNotes("Client# " + getTbNumber(patient, encounter) + ", has missing " + identifierTitle + " identifier");
```
Change this message to "Patient has a missing XXX identifier" - since at times there are no identifiers so ends up as "Client# , has missing Unit TB Number identifier"

moshonk replied:
Noted and addressed

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,210 @@
+					ruleResult.setNotes("Patient #" + getTbNumber(patient, encounter) + " has no final TB Outcome 9 months after start of treatment");
+				} else {
+					ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getId());
+					ruleResult.setNotes("Patient #" + patient.getId() + " has no final TB Outcome 9 months after start of treatment");
```
Do not use the System generated ID - at least using the OpenMRS ID since that can be searched for in the Find Patient widget

moshonk replied:
Noted and addressed

moshonk commented:
@ssmusoke, I have made all the requested changes

slubwama commented:
@moshonk Travis fails

moshonk commented:
@slubwama, noted. I'll look into it

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,209 @@
+			Patient patient = obs.getEncounter().getPatient();
+			
+			RuleResult<Patient> ruleResult = new RuleResult<>();
+			ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getUuid());
```
@moshonk The action url should go to the form with the duplicated TB number 

Also the message shown is "The District TB Number is used by another patient" - without specifying the TB number

moshonk replied:
@ssmusoke, Noted and fixed.

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -67,14 +67,14 @@
 			RuleResult<Patient> ruleResult = new RuleResult<>();
-			ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getUuid());
-			ruleResult.setNotes("The District TB Number " + getTbNumber(patient, obs.getEncounter()) + " is used by another patient");
+			ruleResult.setActionUrl("htmlformentryui/htmlform/editHtmlFormWithStandardUi.page?patientId=" + patient.getUuid() + "&encounterId=" + obs.getEncounter().getId());
```
@moshonk You are missing the formId which determines which form is to be loaded

ssmusoke replied:
@moshonk Ignore this comment works fine wiithout the formUuid

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -67,14 +67,14 @@
-			ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getUuid());
-			ruleResult.setNotes("The District TB Number " + getTbNumber(patient, obs.getEncounter()) + " is used by another patient");
+			ruleResult.setActionUrl("htmlformentryui/htmlform/editHtmlFormWithStandardUi.page?patientId=" + patient.getUuid() + "&encounterId=" + obs.getEncounter().getId());
+			ruleResult.setNotes("The " + identifierTitle + " is used by another patient");
```
@moshonk the actual number has to be displayed so that a user can double check

ssmusoke replied:
@moshonk Ignore this comment - it works fine

moshonk replied:
@ssmusoke Sorrt,  I actually thought that's what you had asked me to do here. 
![image](https://user-images.githubusercontent.com/1963527/28995633-b925fba8-79f6-11e7-916f-3d68872bd16d.png) 

I have addressed it.

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -98,14 +98,14 @@
 			RuleResult<Patient> ruleResult = new RuleResult<>();
-			ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getUuid());
-			ruleResult.setNotes("The District TB Number " + getTbNumber(patient, obs.getEncounter()) + " is used by another patient");
+			ruleResult.setActionUrl("htmlformentryui/htmlform/editHtmlFormWithStandardUi.page?patientId=" + patient.getUuid() + "&encounterId=" + obs.getEncounter().getId());
```
@moshonk Missing the formUuid

ssmusoke replied:
@moshonk Ignore this comment - this works fine

moshonk replied:
@ssmusoke The formUuid is not necessary when editing an existing entry. It's only required when making a new entry

ssmusoke replied:
@moshonk Seen ...

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,211 @@
+			for (Encounter encounter : patientEncounters) {
+				RuleResult<Patient> ruleResult = new RuleResult<>();
+				ruleResult.setActionUrl("htmlformentryui/htmlform/editHtmlFormWithStandardUi.page?patientId=" + patient.getUuid() + "&encounterId=" + encounter.getId());
+				ruleResult.setNotes("Patient #" + getTbNumber(patient, encounter) + " has a missing " + identifierTitle );
```
@moshonk In this case you cannot have a TB number since its missing - so in this case use the OpenMRS ID

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,211 @@
+			Patient patient = obs.getEncounter().getPatient();
+			RuleResult<Patient> ruleResult = new RuleResult<>();
+			ruleResult.setActionUrl("htmlformentryui/htmlform/editHtmlFormWithStandardUi.page?patientId=" + patient.getUuid() + "&encounterId=" + obs.getEncounter().getId());
+			ruleResult.setNotes("The " + identifierTitle + "[" + getTbNumber(patient, obs.getEncounter()) + "] is used by another patient");
```
@moshonk Looks like the TB number duplicated is not loaded the message that I get is  "The Unit TB Number[] is used by another patient"

moshonk replied:
@ssmusoke.. this has been fixed

ssmusoke replied:
@moshonk 

1. The unit TB number is returned irrespective of what identifier is selected see attachment 

2. The Patient # (OpenMRS ID) not being completed for no identifiers entered 

3. Remove the [] in which the TB identifiers are displayed 
<img width="524" alt="screenshot 2017-08-05 19 13 05" src="https://user-images.githubusercontent.com/689900/28996938-7ad84e2a-7a12-11e7-8278-5a1c013672b9.png">

moshonk replied:
@ssmusoke. The issue has been noted and fixed.

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,211 @@
+			Patient patient = obs.getEncounter().getPatient();
+			RuleResult<Patient> ruleResult = new RuleResult<>();
+			ruleResult.setActionUrl("htmlformentryui/htmlform/editHtmlFormWithStandardUi.page?patientId=" + patient.getUuid() + "&encounterId=" + obs.getEncounter().getId());
+			ruleResult.setNotes("The " + identifierTitle + "[" + getTbNumber(patient, obs.getEncounter(), identifierConceptUuid) + "] is used by another patient");
```
@moshonk This message should say that the TB identifier is used by the same patient across multiple treatment programs ... Also remove the [ ] and space around the TB identifier

moshonk replied:
@ssmusoke Noted and rectified,

ssmusoke commented on api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java:
```diff
@@ -0,0 +1,211 @@
+	 * Patients with similar TB identifiers: duplicated for a single patient,
+	 * @return List<RuleResult<Patient>>
+	 */
+	public List<RuleResult<Patient>> singlePatientWithDuplicateTBNumberAcrossMultipleTreatmentPrograms(String identifierConceptUuid, String identifierTitle) {
```
@moshonk This method should have a filter for the same patient otherwise it duplicates the functionality of multiplePatientsWithTheSameTBIdentifiers. Also ensure that multiplePatientsWithTheSameTBIdentifiers does not return the same patient

ssmusoke replied:
@moshonk The filter still does not work ... 

1. A patient with the a duplicated identifier with another patient works fine

2. However the rule above is shown as an error for a patient with the same identifier across multiple programs - look into this particular query as the duplicated obs have to be for the same patient

moshonk replied:
@ssmusoke, noted and fixed.

ssmusoke commented:
@moshonk This has not been fixed... Do the following:

1. Create a patient and create 2 separate TB summary forms 2 years apart with the same identifiers. 

2. Create 2 more patients with duplicate TB identifiers 

#1 will fail the validation for patient across multiple programs, but will also fail on the second one where the identifiers are seen by multiple patients

moshonk commented:
@ssmusoke, noted. This was a tricky one to nab but I have it addressed now.
